### PR TITLE
More robust bash completions generation

### PR DESCRIPTION
**Pull request by bmorledge-hampton19**
- Implemented a new bash completions script generator which constructs completions by traversing the current list of arguments from the beginning, recursing into subparsers where necessary and paying attention to nargs in order to determine which positional arguments are active.
- Removed _shtab_replace_hyphen() function from generated completions script as it seems to have been made obsolete by _shtab_replace_nonword()
- DISCLAIMER: This is my first foray into complex bash scripting, completions, and collaboration through github, pull requests, etc.  I honestly believe the changes I have made are an improvement, especially in resolving #11, but I'll admit that I'm not crystal clear on all the inner workings of shtab at the moment.  I am more than willing to revise this commit and contribute further to the project as necessary.

### CODE REVIEW

**Comment by casperdcl** [comment]
- bash: more robust completion (#38 <- #37)

fixes #37

**Comment by casperdcl** [comment]
what bash version are you using? I'm getting syntax errors.

**Comment by bmorledge-hampton19** [comment]
I tested the script using GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu).  Maybe the differences are between versions 4 and 5?  What exactly was causing the syntax errors?

**Comment by bmorledge-hampton19** [comment]
Ah.  I see what the problem is.  I didn't realize that '+' was a valid argument to nargs.  Here are the lines that need to be changed:
```
if [[ "$current_action_nargs" != "*" ]] && \\
      (( $word_index-$current_action_index+$optional_action_offset \\
         >= $current_action_nargs )); then
```
All that needs to be added is an additional check to ensure that `$currenct_action_nargs` does not equal "+"

**Comment by casperdcl** [comment]
`?`, `*`, and `+` are all valid

File: shtab/__init__.py [hunk 1]
```diff
@@ -275,85 +399,133 @@ def complete_bash(
...
+    # or because no more input is expected from the current action,
+    # indicating that the next positional action can fill in here.
+    if [[ "${current_option_strings[@]}" =~ "${this_word}" ]]; then
+      _set_new_action $this_word false
+      optional_action_offset=0
+    else
+      optional_action_offset=1
+    fi
+
+    if [[ "$current_action_nargs" != "*" ]] && \\
```
Comment (casperdcl) [inline shtab/__init__.py:491]:
```suggestion
    if [[ "$current_action_nargs" != "?" ]] && \\
       [[ "$current_action_nargs" != "*" ]] && \\
       [[ "$current_action_nargs" != "+" ]] && \\
```

should there be an `else` clause too though?

Comment (bmorledge-hampton19) [reply shtab/__init__.py:491]: Got it.  The `?` case is actually accounted for already, as it is parsed earlier in the python script to the default "nargs = 1" case.  As a result, the `if [[ "$current_action_nargs" != "?" ]] && \\` line is redundant.

I don't think an else clause is necessary here.  The if statement evaluates to true when the number of arguments for a given action has reached nargs.  When this occurs, the script moves on to the next positional action.  In the case where nargs has not yet been reached (or will never be reached, in the case of `*` or `+`) the script only needs to _not_ move on to the next action.  This is equivalent to just not executing the code within the if statement, so no else statement is necessary.  At least, I _think_ that's how it should work.  Please let me know if my logic is faulty here.

Comment (bmorledge-hampton19) [reply shtab/__init__.py:491]: Hmmm...  I just had another thought:  Do we need to account for the case where arguments to a positional action are split up by an optional action?  Is this even allowed by argparse? 
e.g. `command arg_1_to_pos_action_1 --opt-action arg_to_opt-action arg_2_to_pos_action_1 arg_to_pos_action_2`
Currently, this isn't expected or handled properly by the current script.

Comment (casperdcl) [reply shtab/__init__.py:491]: `argparse` does indeed allow `<pos> --opt <pos>` :)

File: shtab/__init__.py [hunk 2]
```diff
@@ -177,63 +167,191 @@ def get_optional_actions(parser):
...
+                        u"{}_pos_{}_choices='{}'".format(
+                            prefix, i, " ".join(this_positional_choices)
+                        )
+                    )
+
+            # Lastly, check for information on nargs.
+            # (Skipping any nargs that are 1, None, '?', or "A..." as these are
+            # considered default.)
+            # NOTE: "A..." seems to be associated with calls to subparsers.
+            # I'm not sure exactly what it means...
```
Comment (casperdcl) [inline shtab/__init__.py:268]:
```suggestion
            # Lastly, skip default `nargs` values
```

Comment (casperdcl) [inline shtab/__init__.py:268]:
```suggestion
            # Lastly, skip default `nargs` values
```

File: shtab/__init__.py [hunk 3]
```diff
@@ -275,85 +399,134 @@ def complete_bash(
...
+    # indicating that the next positional action can fill in here.
+    if [[ "${current_option_strings[@]}" =~ "${this_word}" ]]; then
+      _set_new_action $this_word false
+      optional_action_offset=0
+    else
+      optional_action_offset=1
+    fi
+
+    if [[ "$current_action_nargs" != "*" ]] && \\
+       [[ "$current_action_nargs" != "+" ]] && \\
```
Comment (casperdcl) [inline shtab/__init__.py:492]:
```suggestion
       [[ "$current_action_nargs" != "+" ]] && \\
       [[ "$current_action_nargs" != *"..." ]] && \\
```

**Review by casperdcl — CHANGES_REQUESTED** [review]
subparsers & commands need fixing. Otherwise:

```
dvc metrics <TAB>
bash: ((: 1-1+1          >= A... : syntax error: invalid arithmetic operator (error token is "... ")
```

File: shtab/__init__.py (hunk 3 above)
Comment (bmorledge-hampton19) [reply shtab/__init__.py:492]: This looks good to me!  Thank you for resolving this.

**Comment by casperdcl** [comment]
@bmorledge-hampton19 I've just released a new version (v1.3.9) and rebased this PR and found another bug.

Custom `.complete` functionality sometimes doesn't work. Namely:

https://github.com/iterative/shtab/blob/4731037873424798401c1958110ef1cf5f3e7979/examples/customcomplete.py#L49

**Comment by bmorledge-hampton19** [comment]
Hmmm... Was this functionality implemented before these changes?  From what I understand, the `.complete = TXT_FILE` statement sets the COMPGEN value to `_shtab_greeter_compgen_TXTFiles`.  If I'm understanding things correctly, this means there should be a function in the generated bash script with this name, similar to:

https://github.com/iterative/shtab/blob/dd9f7fea19bda59bd07808d80b52947852f27c6c/shtab/__init__.py#L389-L391

However, I can't find any such function for `_shtab_greeter_compgen_TXTFiles`.  Does this functionality need to be added, or am I just missing something?

**Comment by casperdcl** [comment]
yes this functionality always existed - provided via a user-specified preamble:

https://github.com/iterative/shtab/blob/4731037873424798401c1958110ef1cf5f3e7979/examples/customcomplete.py#L18-L22

which is injected into the generated bash script... but after this PR the injected function isn't being used.

**Comment by bmorledge-hampton19** [comment]
I'm having difficulties understanding how the custom function gets injected into the bash script.  From what I understand, the function is passed as a variable called `PREAMBLE`

https://github.com/iterative/shtab/blob/dd9f7fea19bda59bd07808d80b52947852f27c6c/examples/customcomplete.py#L43-L45

Then, this preamble becomes part of the `PrintCompleteAction` class which in turn becomes the value associated with the `action` keyword in a new argument for the parser.

https://github.com/iterative/shtab/blob/dd9f7fea19bda59bd07808d80b52947852f27c6c/shtab/__init__.py#L748-L754
https://github.com/iterative/shtab/blob/dd9f7fea19bda59bd07808d80b52947852f27c6c/shtab/__init__.py#L775-L784

When the `PrintCompleteAction` class is called, it begins generating the completions script through the `complete` function and passes the preamble along with it so that it is injected into the completions script.

What I don't understand is how this class gets called in the first place.  It seems to me that all the changes I've made only touch code that runs after the `complete` function has been invoked, so I'm confused as to how the preamble value isn't getting passed to that function as it was before.

In summary:  Where/how is the `PrintCompleteAction` class called in order to pass along the preamble to the `complete_bash` function?

**Comment by casperdcl** [comment]
> so I'm confused as to how the preamble value isn't getting passed to that function as it was before

is it being passed correctly still, that's not the problem.

```bash
python examples/customcomplete.py completion
```

outputs:

```
...
_shtab_customcomplete_process_pos_0_COMPGEN=_shtab_greeter_compgen_TXTFiles
...
# Custom Preamble                                
                                                 
# $1=COMP_WORDS[1]                               
_shtab_greeter_compgen_TXTFiles() {              
  compgen -d -S '/' -- $1  # recurse into subdirs
  compgen -f -X '!*?.txt' -- $1                  
  compgen -f -X '!*?.TXT' -- $1                  
}                                                
                                                 
# End Custom Preamble                            
...
```

The problem is `_shtab_customcomplete_process_pos_0_COMPGEN` isn't being used properly.

**Comment by bmorledge-hampton19** [comment]
Oh!  I didn't realize that the `add_argument_to` function was a way to invoke shtab's functionality through a custom script's CLI.  (Am I interpreting that correctly?)  I was using a direct call to shtab and the parser argument to generate completions for customcomplete.py, which was not adding in the preamble.  

For what it's worth, the functionality of `shtab.add_argument_to` was never clear to me from the readme or examples.  If it's meant to be used by end-users, maybe a more thorough explanation of what it's doing would be useful?  (Or at least a more descriptive comment than `# magic!`)  Just my thoughts on the matter!  Feel free to disregard.

But I digress.  Once I was generating completions for customcomplete.py properly, I found the issue along with a couple others I had missed previously and (hopefully) fixed them.  The details are in the notes for the commit: https://github.com/iterative/shtab/pull/38/commits/2ce11886d351f1dcafcf25e5b7f3fa3c94208146

**Comment by bmorledge-hampton19** [comment]
I'm also realizing that bits of my code are getting flagged by "flake8" for breaking conventions.  I apologize for not being aware of these conventions sooner.  I've added a new commit to address the issues it flagged. https://github.com/iterative/shtab/pull/38/commits/74930474ca768b2bd12dffcb3dc195cf516e0b09

**Comment by casperdcl** [comment]
Thanks! Just fixed all the tests (I think)...

File: shtab/__init__.py [hunk 4]
```diff
@@ -275,85 +381,122 @@ def complete_bash(
...
+  # Generate the completions
+
+  if [[ "${completing_word}" == -* ]]; then
+    # optional argument started: use option strings
+    COMPREPLY=( $(compgen -W "${current_option_strings[*]}" -- "${completing_word}") )
+  else
+    # use choices & compgen
+    COMPREPLY=( $(compgen -W "${current_action_choices}" -- "${completing_word}"; \\
+                  [ -n "${current_action_compgen}" ] \\
+                  && "${current_action_compgen}" "${completing_word}") )
```
Comment (casperdcl) [inline shtab/__init__.py:481]: not too sure about this...

```
[ -n "${current_action_compgen}" ] && "${current_action_compgen}" "${completing_word}"
```

so `current_action_compgen` is a command which takes one positional arg `completing_word`?

Comment (bmorledge-hampton19) [reply shtab/__init__.py:481]: Sorry about the late reply.  Your assessment is correct.  `current_action_compgen` will be a command such as `_shtab_compgen_files()` or `_shtab_compgen_dirs()` which takes one argument to generate possible completions.

Comment (bmorledge-hampton19) [reply shtab/__init__.py:481]: Is that okay?  Are there some custom compgen functions which aren't compatible with this setup?

**Review by casperdcl — APPROVED** [review]
epic stuff!

File: shtab/__init__.py (hunk 4 above)
Comment (casperdcl) [reply shtab/__init__.py:481]: lgtm
